File: Source/Assets/Scripts/custom/BulletShot.js
#pragma strict
// a little script for auto-destroy projectile instances when get out of screen or after some time lapse...

public var LifeTime			: float = 5f;
public var speed			: float = 5f;
public var rotationSpeed	: float = 0;

//private var Fired			: boolean = false;
private var moveDirection 	: Vector3 ;

public var RowSize			: int = 1;		// Input the number of Rows of the sprite sheet 
public var ColumnSize		: int = 1;		// Input the number of columns of the sprite sheet 
public var framesPerSecond 	: int = 12;		// Speed of Sprite animation

public var rowFrameStart	: int = 1;
public var colFrameStart	: int = 1;
public var totalFrames		: int = 1;

function Fire( moveSpeed : Vector3, rotSpeed : float)
{
	moveDirection 	= moveSpeed; 	// == transform.up * speed
	rotationSpeed 	= rotSpeed;
		PlayFrames(rowFrameStart, colFrameStart, totalFrames, Mathf.Sign( moveDirection.x ));
	
	while ( true)
		yield CoUpdate();
//	Fired 			= true;
}

function FireAnimated( moveSpeed : Vector3, rowStart :int, colStart :int, totalframes :int)
{
	moveDirection 	= moveSpeed; 	// == transform.up * speed ( orientation of move + force of impulse )
	rowFrameStart 	= rowStart;
	colFrameStart	= colStart;
	totalFrames		= totalframes;
	
	while ( true)
	{
		CoUpdate();
		PlayFrames(rowFrameStart, colFrameStart, totalFrames, Mathf.Sign( moveDirection.x ));
		yield;
	}
//	Fired 			= true;
}


function CoUpdate () : IEnumerable
{
//	if (Fired)
//	{
    	transform.position += Time.deltaTime * moveDirection;		// moveDirection = orientation * speed ;

		transform.RotateAroundLocal( Vector3.forward, rotationSpeed * Time.deltaTime); // if > 0, Rotate around own z-axis

// 	}
 		Destroy(gameObject, LifeTime);
 		
// 		yield;
}


function PlayFrames( rowFrameStart :int, colFrameStart :int, totalframes :int, flipped :int)
{
 var index 	: int = Time.time * framesPerSecond;							// time control fps
 index = index % totalFrames; 
 
 var size 	: Vector2 = Vector2(  1.0 / ColumnSize , 1.01 / RowSize );	// scale
 
 var u 		: int = index % ColumnSize;
 var v 		: int = index / ColumnSize;
   
 var offset = Vector2 ( ((u + colFrameStart ) * size.x ) , (1.0 - size.y) - ( v + rowFrameStart ) * size.y ); // offset
 
  offset.x = (( offset.x * flipped ) - size.x *  System.Convert.ToByte(flipped < 0)  ) * flipped  ;
  size.x *= flipped ;
  
 renderer.material.mainTextureOffset = offset ;		// texture offset
 renderer.material.mainTextureScale = size  ;		// texture scale
}


//function Start ()
//{
//	TotalTime = Time.time + LifeTime;	
//	moveDirection = transform.up;
//}

//function Update ()
//{
//	if (Fired)
//	{ 
//    transform.position += Time.deltaTime * speed * moveDirection;
//	
//	  rotation  = rotationSpeed * Time.deltaTime;
//		transform.RotateAroundLocal( Vector3.forward, rotation);
//	
//	if ( Mathf.Abs( transform.position.y) > ScreenOffset || Mathf.Abs( transform.position.x) > ScreenOffset )
//	if ( Mathf.Abs( transform.position.y) > ScreenOffset || Time.time > TotalTime )
//  { Destroy(gameObject); }
//}

File: Source/Assets/Scripts/custom/BulletShot.cs
// a little script for auto-destroy projectile instances when get out of screen or after some time lapse...

using UnityEngine;
using System.Collections;

public class BulletShot : MonoBehaviour
{

    public float LifeTime = 5f;
    public float rotationSpeed = 0;	// Add some value for rotation

    public int RowSize = 1;		// Input the number of Rows of the sprite sheet 
    public int ColumnSize = 1;		// Input the number of columns of the sprite sheet 
    public int framesPerSecond = 12;		// Speed of Sprite animation

    public int rowFrameStart = 1;
    public int colFrameStart = 1;
    public int totalFrames = 1;

    private Transform thisTransform;	// own obj's tranform cached
    private Vector3 moveDirection;		// == orientation * speed (force Magnitude )
    private int orientation = 1;


    public void Fire(Vector3 moveSpeed, float rotSpeed)
    {
        thisTransform = transform;
        moveDirection = moveSpeed; 	// == transform.up * speed
        rotationSpeed = rotSpeed;
        orientation = (int)Mathf.Sign(moveDirection.x);
        PlayFrames(rowFrameStart, colFrameStart, totalFrames, orientation);

        //while (true)
        //    yield return new CoUpdate();
        StartCoroutine(CoUpdate());
    }

    public void FireAnimated(Vector3 moveSpeed, int rowStart, int colStart, int totalframes)
    {
        thisTransform = transform;
        moveDirection = moveSpeed; 	// == transform.up * speed ( orientation of move + force of impulse )
        rowFrameStart = rowStart;
        colFrameStart = colStart;
        totalFrames = totalframes;
        orientation = (int)Mathf.Sign(moveDirection.x);

        StartCoroutine(MoveAnimated());
    }

    IEnumerator MoveAnimated()
    {
        StartCoroutine(CoUpdate());

        while (true)
        {
            PlayFrames(rowFrameStart, colFrameStart, totalFrames, orientation);
            yield return 0;
        }
    }


    public void FireBall(Vector3 moveSpeed, int rowStart, int colStart, int totalframes)
    {
        thisTransform = transform;
        moveDirection = moveSpeed; 	// == transform.up * speed ( orientation of move + force of impulse )
        rowFrameStart = rowStart;
        colFrameStart = colStart;
        totalFrames = totalframes;
        orientation = (int)Mathf.Sign(moveDirection.x);

        StartCoroutine(MoveFlamed());
    }

    IEnumerator MoveFlamed()
    {
        StartCoroutine(CoUpdate());

        while (true)
        {
            PlayFrames(rowFrameStart, colFrameStart, totalFrames, orientation);
            thisTransform.localScale = Vector3.one + (Vector3.up * Random.Range(0, 2));
            yield return 0;
        }
    }

   //void Start()
   // {
   //    if ( totalFrames > 1 )
   //        FireAnimated(Vector3.zero, rowFrameStart, colFrameStart, totalFrames);
   // }




    public void FireBoomerang(Vector3 moveSpeed, int rowStart, int colStart, int totalframes) // Hat throw with a boomerang Fx
    {
        thisTransform = transform;
        moveDirection = moveSpeed; 	// == transform.up * speed ( orientation of move + force of impulse )
        rowFrameStart = rowStart;
        colFrameStart = colStart;
        totalFrames = totalframes;
        orientation = (int)Mathf.Sign(moveDirection.x);

        StartCoroutine(MoveBoomerang(moveSpeed));
    }

    IEnumerator MoveBoomerang(Vector3 moveSpeed)
    {
        StartCoroutine( CoUpdate() );

        while (true)
        {
            moveDirection.x += Mathf.Clamp(Time.deltaTime * 12.0f * -orientation,
                                                      moveSpeed.x * -orientation,
                                                      moveSpeed.x * orientation);
            PlayFrames(rowFrameStart, colFrameStart, totalFrames, (int)Mathf.Sign(moveDirection.x));
            yield return 0;
        }
    }


    IEnumerator CoUpdate()
    {
        while (true)
        {
            thisTransform.position += Time.deltaTime * moveDirection;							// moveDirection == orientation * speed ;

            thisTransform.Rotate(Vector3.forward, rotationSpeed * Time.deltaTime); 	// if > 0, Rotate around own z-axis

            Destroy(gameObject, LifeTime);

            yield return 0;
        }
    }
    void PlayFrames(int rowFrameStart, int colFrameStart, int totalframes, int flipped)
    {
        int index = (int)(Time.time * framesPerSecond);							// time control fps
        index = index % totalFrames;

        Vector2 size = new Vector2(1.0f / ColumnSize, 1.01f / RowSize);	// scale

        int u = index % ColumnSize;
        int v = index / ColumnSize;

        Vector2 offset = new Vector2(((u + colFrameStart) * size.x), (1.0f - size.y) - (v + rowFrameStart) * size.y); // offset

        offset.x = ((offset.x * flipped) - size.x * System.Convert.ToByte(flipped < 0)) * flipped;
        size.x *= flipped;

        renderer.material.mainTextureOffset = offset;		// texture offset
        renderer.material.mainTextureScale = size;		// texture scale
    }

}